# Vibe Coding: Real-World Data Cleaning Challenge

## The Mission

You're a Data Analyst at **TechSalary Insights**. Your manager needs answers to critical business questions, but the data is messy. Your job is to clean it and provide accurate insights.

**The catch:** You must figure out how to clean the data yourself. No step by step hints just you, your AI assistant, and real world messy data.

---

## The Dataset: Ask A Manager Salary Survey 2021

**Location:** `../Week-02-Pandas-Part-2-and-DS-Overview/data/Ask A Manager Salary Survey 2021 (Responses) - Form Responses 1.tsv`

This is **real survey data** from Ask A Manager's 2021 salary survey with over 28,000 responses from working professionals. The data comes from this survey: https://www.askamanager.org/2021/04/how-much-money-do-you-make-4.html

**Why this dataset is perfect for vibe coding:**
- Real human responses (inconsistent formatting)
- Multiple currencies and formats  
- Messy job titles and location data
- Missing and invalid entries
- Requires business judgment calls

---

## Your Business Questions

Answer these **exact questions** with clean data. There's only one correct answer for each:

### Core Questions (Required):
1. **What is the median salary for Software Engineers in the United States?** 
2. **Which US state has the highest average salary for tech workers?**
3. **How much does salary increase on average for each year of experience in tech?**
4. **Which industry (besides tech) has the highest median salary?**

### Bonus Questions (If time permits):
5. **What's the salary gap between men and women in tech roles?**
6. **Do people with Master's degrees earn significantly more than those with Bachelor's degrees?**

**Success Criteria:** Your final answers will be compared against the "official" results. Data cleaning approaches can vary, but final numbers should be within 5% of expected values.


---
# Your Work Starts Here

## Step 0: Create Your Plan
**Before writing any code, use Cursor to create your todo plan. Then paste it here:**

## My Data Cleaning Plan

- [ ] Load the TSV dataset into a Pandas DataFrame and inspect columns
- [ ] Explore missing values, data types, and unique values for key columns
- [ ] Clean and standardize the **salary** column (remove symbols, commas, text, convert to numeric USD)
- [ ] Handle non-USD currencies (decide whether to drop or convert)
- [ ] Remove extreme outliers in salary data (<$100 or >$1M)
- [ ] Normalize **job titles** and flag “Software Engineer” and similar variations
- [ ] Extract and standardize **location** into country and US state (map abbreviations/full names)
- [ ] Filter out non-US rows for US-specific questions
- [ ] Convert **years of experience** into numeric form (handle ranges, “less than 1”, etc.)
- [ ] Clean and group **industry** column (e.g., combine synonyms, mark “tech” industries)
- [ ] Standardize **gender** categories (man, woman, nonbinary, etc.)
- [ ] Standardize **education** categories (Bachelor’s, Master’s, PhD, etc.)
- [ ] Compute median salary for US Software Engineers
- [ ] Compute state with highest average tech salary
- [ ] Estimate average salary increase per year of experience in tech (regression or slope)
- [ ] Find highest median salary industry (excluding tech)
- [ ] Bonus: Compute salary gap between men and women in tech
- [ ] Bonus: Compare median salaries of Master’s vs Bachelor’s degrees

## Step 1: Data Loading and Exploration

Start here! Load the dataset and get familiar with what you're working with.


In [77]:
# Import necessary libraries
import pandas as pd
import numpy as np
import re
#import matplotlib.pyplot as plt
#import seaborn as sns

# Step 1: Load the dataset

data_file_path = "A:\Desktop\CTP\Fall-2025\ds-fall-2025-fri-0630\Week-02-Pandas-Part-2-and-DS-Overview\data\Ask A Manager Salary Survey 2021 (Responses) - Form Responses 1.tsv"
# Load the dataset
df = pd.read_csv(data_file_path, sep='\t')\

# Step 2: Explore Data
# Check missing values, column names, and data types.

df.info()
df.head(30)


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 28062 entries, 0 to 28061
Data columns (total 18 columns):
 #   Column                                                                                                                                                                                                                                Non-Null Count  Dtype  
---  ------                                                                                                                                                                                                                                --------------  -----  
 0   Timestamp                                                                                                                                                                                                                             28062 non-null  object 
 1   How old are you?                                                                                             

<>:10: SyntaxWarning: invalid escape sequence '\D'
<>:10: SyntaxWarning: invalid escape sequence '\D'
C:\Users\eflor\AppData\Local\Temp\ipykernel_13708\851499286.py:10: SyntaxWarning: invalid escape sequence '\D'
  data_file_path = "A:\Desktop\CTP\Fall-2025\ds-fall-2025-fri-0630\Week-02-Pandas-Part-2-and-DS-Overview\data\Ask A Manager Salary Survey 2021 (Responses) - Form Responses 1.tsv"


Timestamp How old are you?         What industry do you work in?  \
0   4/27/2021 11:02:10            25-34          Education (Higher Education)   
1   4/27/2021 11:02:22            25-34                     Computing or Tech   
2   4/27/2021 11:02:38            25-34         Accounting, Banking & Finance   
3   4/27/2021 11:02:41            25-34                            Nonprofits   
4   4/27/2021 11:02:42            25-34         Accounting, Banking & Finance   
5   4/27/2021 11:02:46            25-34          Education (Higher Education)   
6   4/27/2021 11:02:51            25-34                            Publishing   
7   4/27/2021 11:03:00            25-34         Education (Primary/Secondary)   
8   4/27/2021 11:03:01            45-54                     Computing or Tech   
9   4/27/2021 11:03:02            35-44         Accounting, Banking & Finance   
10  4/27/2021 11:03:03            25-34                            Nonprofits   
11  4/27/2021 11:03:07            35-44          Education (Higher Education)   
12  4/27/2021 11:03:09            35-44         Accounting, Banking & Finance   
13  4/27/2021 11:03:10            25-34                                   Law   
14  4/27/2021 11:03:11            18-24                           Health care   
15  4/27/2021 11:03:11            35-44        Utilities & Telecommunications   
16  4/27/2021 11:03:13            35-44                Business or Consulting   
17  4/27/2021 11:03:19            45-54                          Art & Design   
18  4/27/2021 11:03:21            35-44                Business or Consulting   
19  4/27/2021 11:03:23            35-44          Education (Higher Education)   
20  4/27/2021 11:03:24            25-34                           Health care   
21  4/27/2021 11:03:24            35-44                            Nonprofits   
22  4/27/2021 11:03:27            35-44                            Nonprofits   
23  4/27/2021 11:03:29            35-44  Government and Public Administration   
24  4/27/2021 11:03:31            25-34                        Public Library   
25  4/27/2021 11:03:34            35-44          Education (Higher Education)   
26  4/27/2021 11:03:38            25-34                            Nonprofits   
27  4/27/2021 11:03:40            35-44                                   Law   
28  4/27/2021 11:03:41            35-44  Government and Public Administration   
29  4/27/2021 11:03:43            25-34          Engineering or Manufacturing   

                                            Job title  \
0                  Research and Instruction Librarian   
1            Change & Internal Communications Manager   
2                                Marketing Specialist   
3                                     Program Manager   
4                                  Accounting Manager   
5                      Scholarly Publishing Librarian   
6                                Publishing Assistant   
7                                           Librarian   
8                                     Systems Analyst   
9                                   Senior Accountant   
10                                     Office Manager   
11  Deputy Title IX Coordinator/ Assistant Directo...   
12                    Manager of Information Services   
13                           Legal Aid Staff Attorney   
14                           Patient care coordinator   
15                  Quality And Compliance Specialist   
16                                Executive Assistant   
17                                   graphic designer   
18                                     Senior Manager   
19            Assistant Director of Academic Advising   
20                            Data Programmer Analyst   
21            Program Coordinator & Assistant Editor    
22                                      Event Planner   
23                                         Researcher   
24                                    Teen Librarian    
25                 

## Step 2: Data Cleaning


In [82]:
#Step 3: Data Cleaning

# Columns names
rename_map = {
    "Timestamp": "timestamp",
    "How old are you?": "age",
    "What industry do you work in?": "industry",
    "Job title": "job_title",
    "If your job title needs additional context, please clarify here:": "job_context",
    "What is your annual salary? (You'll indicate the currency in a later question. If you are part-time or hourly, please enter an annualized equivalent -- what you would earn if you worked the job 40 hours a week, 52 weeks a year.)": "salary",
    "How much additional monetary compensation do you get, if any (for example, bonuses or overtime in an average year)? Please only include monetary compensation here, not the value of benefits.": "additional_comp",
    "Please indicate the currency": "currency",
    'If "Other," please indicate the currency here:': 'currency_other',
    "If your income needs additional context, please provide it here:": "income_context",
    "What country do you work in?": "country",
    "If you're in the U.S., what state do you work in?": "state",
    "What city do you work in?": "city",
    "How many years of professional work experience do you have overall?": "overall_exp",
    "How many years of professional work experience do you have in your field?": "field_exp",
    "What is your highest level of education completed?": "education",
    "What is your gender?": "gender",
    "What is your race? (Choose all that apply.)": "race"
}

df = df.rename(columns=rename_map)

## Step 3: Clean Salary Column
# Strip `$`, `,`, and other text from salary  
# Convert to numeric  
# Filter out unrealistic values (<100, >1,000,000)

df['salary'] = (df['salary']
    .astype("str")                               # make sure all values are strings
    .str.replace(r'[^0-9.]', '', regex=True)   # keep only digits + decimal point
)

#Convert the cleaned strings to actual numbers (floats)
df['salary'] = pd.to_numeric(df['salary'], errors='coerce')

# Step 3d: Filter out unrealistic values
df = df[df['salary'].between(100, 1_000_000)]

#Keep only rows where the currency is 'USD'
df = df[df['currency'] == 'USD']

# Fix common variants of US
df['country'] = df['country'].replace({
    'us': 'united States',
    'united states': 'United States',
    'US': 'united States',
    'USA': 'united states'
})

## Step 7: Range Columns to midpoint
#These are categorical ranges like "2 - 4 years" or "11 - 20 years".
#We’ll convert them to numeric by taking the midpoint.

def age_to_midpoint(value: str) -> float:
    """Convert age ranges like '25-34' into numeric midpoints."""
    if pd.isna(value):
        return None
    
    val = str(value).lower().strip()
    
    # Extract all numbers
    match = re.findall(r'\d+', val)  # <-- FIX: '\d+' not '(\d)'
    
    if len(match) == 2:   # e.g., "25-34"
        return (int(match[0]) + int(match[1])) / 2
    elif "or more" in val and len(match) == 1:  # e.g., "65 or more"
        return float(match[0]) + 5   # assume midpoint ~70
    elif len(match) == 1:
        return float(match[0])
    
    return None

def exp_to_midpoint(value: str) -> float:
    """Convert experience strings like '2-4 years' or '21 years or more' into numeric years."""
    if pd.isna(value):
        return None
    
    val = str(value).lower().strip()
    
    if "less than 1" in val:
        return 0.5
    
    match = re.findall(r'\d+', val)  # FIX: use full numbers
    
    if len(match) == 2:
        return (int(match[0]) + int(match[1])) / 2
    elif "or more" in val and len(match) == 1:
        return float(match[0])  # just take the lower bound
    elif len(match) == 1:
        return float(match[0])
    
    return None

# Apply to columns
df['age'] = df['age'].apply(age_to_midpoint)
df['overall_exp'] = df['overall_exp'].apply(exp_to_midpoint)
df['field_exp'] = df['field_exp'].apply(exp_to_midpoint)

#dorp nan salary
df = df.dropna(subset=['salary'])

# Fill in categorical NaNs with 'Unknown'
for col in ['industry', 'education', 'gender', 'race','job_context','state']:
    df[col] = df[col].fillna('Unknown')


df.head(20)

timestamp   age                              industry  \
0   4/27/2021 11:02:10  29.5          Education (Higher Education)   
2   4/27/2021 11:02:38  29.5         Accounting, Banking & Finance   
3   4/27/2021 11:02:41  29.5                            Nonprofits   
4   4/27/2021 11:02:42  29.5         Accounting, Banking & Finance   
5   4/27/2021 11:02:46  29.5          Education (Higher Education)   
6   4/27/2021 11:02:51  29.5                            Publishing   
7   4/27/2021 11:03:00  29.5         Education (Primary/Secondary)   
8   4/27/2021 11:03:01  49.5                     Computing or Tech   
9   4/27/2021 11:03:02  39.5         Accounting, Banking & Finance   
10  4/27/2021 11:03:03  29.5                            Nonprofits   
11  4/27/2021 11:03:07  39.5          Education (Higher Education)   
12  4/27/2021 11:03:09  39.5         Accounting, Banking & Finance   
13  4/27/2021 11:03:10  29.5                                   Law   
16  4/27/2021 11:03:13  39.5                Business or Consulting   
17  4/27/2021 11:03:19  49.5                          Art & Design   
18  4/27/2021 11:03:21  39.5                Business or Consulting   
19  4/27/2021 11:03:23  39.5          Education (Higher Education)   
20  4/27/2021 11:03:24  29.5                           Health care   
21  4/27/2021 11:03:24  39.5                            Nonprofits   
23  4/27/2021 11:03:29  39.5  Government and Public Administration   

                                            job_title  \
0                  Research and Instruction Librarian   
2                                Marketing Specialist   
3                                     Program Manager   
4                                  Accounting Manager   
5                      Scholarly Publishing Librarian   
6                                Publishing Assistant   
7                                           Librarian   
8                                     Systems Analyst   
9                                   Senior Accountant   
10                                     Office Manager   
11  Deputy Title IX Coordinator/ Assistant Directo...   
12                    Manager of Information Services   
13                           Legal Aid Staff Attorney   
16                                Executive Assistant   
17                                   graphic designer   
18                                     Senior Manager   
19            Assistant Director of Academic Advising   
20                            Data Programmer Analyst   
21            Program Coordinator & Assistant Editor    
23                                         Researcher   

                     job_context  salary  additional_comp currency  \
0                        Unknown   55000              0.0      USD   
2                        Unknown   34000              NaN      USD   
3                        Unknown   62000           3000.0      USD   
4                        Unknown   60000           7000.0      USD   
5                        Unknown   62000              NaN      USD   
6                        Unknown   33000           2000.0      USD   
7                High school, FT   50000              NaN      USD   
8   Data developer/ETL Developer  112000          10000.0      USD   
9                        Unknown   45000              0.0      USD   
10                       Unknown   47500              0.0      USD   
11                       Unknown   62000              0.0      USD   
12                       Unknown  100000              0.0      USD   
13           non-profit law firm   52000              0.0      USD   
16                       Unknown   85000           5000.0      USD   
17                       Unknown   59000              NaN      USD   
18                       Unknown   98000           1000.0      USD   
19                       Unknown   54000              NaN      USD   
20                       Unknown   74000              0.0      USD   
21            

## Step 3: Business Questions Analysis

Now answer those important business questions!


In [83]:
# Question 1: What is the median salary for Software Engineers in the United States?

# Step 1: Filter US-only
us_df = df[df['country'].str.lower() == 'united states']

# Step 2: Define SWE detector
def is_software_engineer(title: str) -> bool:
    if pd.isna(title):
        return False
    title = str(title).lower()
    keywords = [
        "software engineer",
        "swe",
        "software developer",
        "programmer",
        "engineer (software)"
    ]
    return any(kw in title for kw in keywords)

# Step 3: Flag SWE jobs
us_df['is_swe'] = us_df['job_title'].apply(is_software_engineer)

# Step 4: Filter to SWE only
swe_us = us_df[us_df['is_swe']]

# Step 5: Median salary
median_salary_swe = swe_us['salary'].median()

print(f"💰 Median salary for Software Engineers in the US: ${median_salary_swe:,.0f}")
print("Number of SWE respondents:", len(swe_us))

💰 Median salary for Software Engineers in the US: $140,000
Number of SWE respondents: 824


C:\Users\eflor\AppData\Local\Temp\ipykernel_13708\432605178.py:21: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  us_df['is_swe'] = us_df['job_title'].apply(is_software_engineer)


In [84]:
# Question 2: Which US state has the highest average salary for tech workers?

# Step 1: Filter to US only
us_df = df[df['country'].str.lower() == 'united states']

# Step 2: Define tech worker detector (based on industry for now)
def is_tech_worker(industry: str, job_title: str) -> bool:
    """
    Identify if a worker is in tech by checking industry or job title.
    """
    if pd.isna(industry) and pd.isna(job_title):
        return False
    
    text = f"{industry} {job_title}".lower()
    
    tech_keywords = [
        "tech", "computing", "software", "information technology",
        "it", "developer", "engineer", "data scientist", "programmer"
    ]
    
    return any(kw in text for kw in tech_keywords)

# Step 3: Flag tech workers
us_df['is_tech'] = us_df.apply(lambda row: is_tech_worker(row['industry'], row['job_title']), axis=1)

# Step 4: Filter tech workers
tech_us = us_df[us_df['is_tech']]

# Step 5: Group by state and compute average salary
state_avg_salary = tech_us.groupby('state')['salary'].mean().sort_values(ascending=False)
# Step 6: Find the state with the highest average salary
top_state = state_avg_salary.idxmax()
top_salary = state_avg_salary.max()

print("💻 State with highest average tech salary:", top_state)
print(f"💰 Average salary: ${top_salary:,.0f}")


💻 State with highest average tech salary: Colorado, Nevada
💰 Average salary: $190,000


C:\Users\eflor\AppData\Local\Temp\ipykernel_13708\1897108981.py:24: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  us_df['is_tech'] = us_df.apply(lambda row: is_tech_worker(row['industry'], row['job_title']), axis=1)


In [85]:
# Question 3: How much does salary increase on average for each year of experience in tech?
from sklearn.linear_model import LinearRegression
import numpy as np

# Step 1: Filter US tech workers
us_df = df[df['country'].str.lower() == 'united states']

# Reuse our tech worker flag
def is_tech_worker(industry: str, job_title: str) -> bool:
    if pd.isna(industry) and pd.isna(job_title):
        return False
    text = f"{industry} {job_title}".lower()
    tech_keywords = [
        "tech", "computing", "software", "information technology",
        "it", "developer", "engineer", "data scientist", "programmer"
    ]
    return any(kw in text for kw in tech_keywords)

us_df['is_tech'] = us_df.apply(lambda row: is_tech_worker(row['industry'], row['job_title']), axis=1)

tech_df = us_df[us_df['is_tech']].dropna(subset=['salary', 'overall_exp'])

# Step 2: Prepare features (X) and target (y)
X = tech_df[['overall_exp']]   # years of experience
y = tech_df['salary']          # salary

# Step 3: Fit linear regression
model = LinearRegression()
model.fit(X, y)

# Step 4: Get slope (increase per year)
increase_per_year = model.coef_[0]

print(f"📈 On average, salary increases by ${increase_per_year:,.0f} per year of experience in tech.")

📈 On average, salary increases by $1,447 per year of experience in tech.


C:\Users\eflor\AppData\Local\Temp\ipykernel_13708\1586186880.py:19: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  us_df['is_tech'] = us_df.apply(lambda row: is_tech_worker(row['industry'], row['job_title']), axis=1)


In [86]:

# Question 4: What percentage of respondents work remotely vs. in-office?
def classify_work_mode(text: str) -> str:
    """Classify work mode from free-text context."""
    if pd.isna(text):
        return "Unknown"
    t = text.lower()
    if "remote" in t or "wfh" in t or "work from home" in t or "telecommute" in t:
        return "Remote"
    if "hybrid" in t:
        return "Hybrid"
    if "in office" in t or "on site" in t or "onsite" in t:
        return "In-Office"
    return "Unknown"

# Apply to both job_context and income_context (combine signals)
df['work_mode'] = df[['job_context', 'income_context']].astype(str).apply(
    lambda row: classify_work_mode(" ".join(row)), axis=1
)

# Count percentages
work_mode_counts = df['work_mode'].value_counts(normalize=True) * 100
print(work_mode_counts)

work_mode
Unknown      99.729625
Remote        0.197416
Hybrid        0.042917
In-Office     0.030042
Name: proportion, dtype: float64


In [89]:
# Question 5: Which industry (besides tech) has the highest median salary?

# Step 1: Exclude tech
non_tech = df[~df['industry'].str.lower().str.contains("tech")]

# Step 2: Group by industry and compute median salary
industry_median = non_tech.groupby('industry')['salary'].median().sort_values(ascending=False)

# Step 3: Get the top industry
top_industry = industry_median.index[0]
top_salary = industry_median.iloc[0]

print("🏆 Highest median salary (non-tech industry):", top_industry)
print(f"💰 Median salary: ${top_salary:,.0f}")

# Optional: show top 5 industries
print("\nTop 5 non-tech industries by median salary:")
print(industry_median.head(5))

🏆 Highest median salary (non-tech industry): Commercial Building Material Distribution
💰 Median salary: $400,000

Top 5 non-tech industries by median salary:
industry
Commercial Building Material Distribution                         400000.0
Sports                                                            300000.0
Corporate Training                                                280000.0
Energy (oil & gas & associated products, renewable power, etc)    253300.0
Multilateral Organisation                                         250000.0
Name: salary, dtype: float64


In [97]:
# Bonus Questions:
# Question 6: What's the salary gap between men and women in similar roles?


# Step 1: Filter to US respondents
us_df = df[df['country'].str.lower() == 'united states']

# Step 2: Normalize gender
us_df['gender'] = us_df['gender'].str.strip().str.title()

# Step 3: Keep only Man/Woman (drop nonbinary/unknown for this specific Q)
gender_df = us_df[us_df['gender'].isin(['Man', 'Woman'])]

# Step 4: Group by job title and gender, compute median salary
role_gender_salary = gender_df.groupby(['job_title', 'gender'])['salary'].median().unstack()

# Step 5: Compute gap (male median - female median)
role_gender_salary['gap'] = role_gender_salary['Man'] - role_gender_salary['Woman']

# Step 6: Average gap across all roles
avg_gap = role_gender_salary['gap'].mean()

print(f"📊 On average, men earn about ${avg_gap:,.0f} more than women in similar roles (median-based).")

# Optional: show the top 10 roles with the biggest gaps
print("\nTop 10 roles with biggest male-female pay gaps:")
print(role_gender_salary.sort_values('gap', ascending=False).head(10))

# Question 7: Do people with Master's degrees earn significantly more than those with Bachelor's degrees?


import scipy.stats as stats

# Step 1: Normalize education
df['education'] = df['education'].str.strip().str.title()

# Map education into categories
edu_map = {
    "College Degree": "Bachelors",
    "Bachelor'S Degree": "Bachelors",
    "Master'S Degree": "Masters"
}
df['edu_group'] = df['education'].map(edu_map)

# Step 2: Filter to only Bachelor's and Master's
edu_df = df[df['edu_group'].isin(['Bachelors', 'Masters'])]

# Step 3: Median salaries
median_bach = edu_df[edu_df['edu_group'] == 'Bachelors']['salary'].median()
median_mast = edu_df[edu_df['edu_group'] == 'Masters']['salary'].median()

print(f"🎓 Median salary (Bachelor's): ${median_bach:,.0f}")
print(f"🎓 Median salary (Master's): ${median_mast:,.0f}")

# Step 4: Statistical test (t-test)
bach_salaries = edu_df[edu_df['edu_group'] == 'Bachelors']['salary']
mast_salaries = edu_df[edu_df['edu_group'] == 'Masters']['salary']

t_stat, p_val = stats.ttest_ind(mast_salaries, bach_salaries, equal_var=False)

print("\n📊 T-test Results:")
print(f"t-statistic: {t_stat:.2f}, p-value: {p_val:.4f}")

if p_val < 0.05:
    print("✅ Yes, Master's degree holders earn significantly more than Bachelor's degree holders.")
else:
    print("❌ No significant difference between Master's and Bachelor's degree holders.")
# Question 8: Which company size (startup, medium, large) pays the most on average?



📊 On average, men earn about $8,274 more than women in similar roles (median-based).

Top 10 roles with biggest male-female pay gaps:
gender                       Man     Woman       gap
job_title                                           
Program Associate       510000.0   63000.0  447000.0
Managing director       390000.0   93000.0  297000.0
Design Strategist       250000.0   85000.0  165000.0
VP Engineering          275000.0  115000.0  160000.0
Grants Director         207500.0   54000.0  153500.0
Head of School          210000.0   61000.0  149000.0
Director of Education   205000.0   67000.0  138000.0
Deputy General Counsel  300000.0  164000.0  136000.0
Sales Director          179000.0   46000.0  133000.0
Project Director        195000.0   65000.0  130000.0
🎓 Median salary (Bachelor's): $75,000
🎓 Median salary (Master's): $80,000

📊 T-test Results:
t-statistic: 6.34, p-value: 0.0000
✅ Yes, Master's degree holders earn significantly more than Bachelor's degree holders.


C:\Users\eflor\AppData\Local\Temp\ipykernel_13708\154261645.py:9: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  us_df['gender'] = us_df['gender'].str.strip().str.title()


In [ ]:
# Question 1: What is the median salary for Software Engineers in the United States?

# Step 1: Filter US-only
us_df = df[df['country'].str.lower() == 'united states']

# Step 2: Define SWE detector
def is_software_engineer(title: str) -> bool:
    if pd.isna(title):
        return False
    title = str(title).lower()
    keywords = [
        "software engineer",
        "swe",
        "software developer",
        "programmer",
        "engineer (software)"
    ]
    return any(kw in title for kw in keywords)

# Step 3: Flag SWE jobs
us_df['is_swe'] = us_df['job_title'].apply(is_software_engineer)

# Step 4: Filter to SWE only
swe_us = us_df[us_df['is_swe']]

# Step 5: Median salary
median_salary_swe = swe_us['salary'].median()

print(f"💰 Median salary for Software Engineers in the US: ${median_salary_swe:,.0f}")
print("Number of SWE respondents:", len(swe_us))

💰 Median salary for Software Engineers in the US: $140,000
Number of SWE respondents: 824


C:\Users\eflor\AppData\Local\Temp\ipykernel_13708\432605178.py:21: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  us_df['is_swe'] = us_df['job_title'].apply(is_software_engineer)


## Final Summary

**Summarize your findings here:**

1. **Median salary for Software Engineers in US:** US: $140,000
2. **Highest paying US state for tech:** Colorado, Nevada
3. **Salary increase per year of experience:**  $1,447 
4. **Remote vs office percentage:** Remote: 0.197416  Hybrid: 0.042917 In-Office: 0.030042
5. **Highest paying non-tech industry:** Commercial Building Material Distribution

**Key insights:**
- Cleaning real survey data takes multiple steps (renaming columns, handling NaN, cleaning salaries, standardizing experience/age).
- Many business questions (like gender pay gap or education impact) can be answered once the dataset is tidy and consistent.
- Even advanced topics like regression, calculus, and linear algebra become easier to understand when applied to real data.

**Challenges faced:**
- Understanding how to detect job titles like “Software Engineer” with code. → Solved it by learning how any(kw in title for kw in keywords) works and testing it on samples.
- Challenge 2: Cleaning ranges like “25–34” or “21 years or more.” → Solved it by writing custom functions (age_to_midpoint, exp_to_midpoint) and fixing regex to capture full numbers.

**What you learned about vibe coding:**
- Learning 1: Vibe coding lets me learn while I build — no rigid instructions, just exploration and problem solving.
- Learning 2: I realized data cleaning and analysis can be smooth and even fun when I treat it like a conversation with the dataset.
- Learning 3: I felt confident because everything flowed quickly; I wasn’t stuck, and I could understand the concepts better in context.